jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codewhisperer.telemetry
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.editor.event.DocumentEvent
+import com.intellij.util.Alarm
+import com.intellij.util.AlarmFactory
+import org.jetbrains.annotations.TestOnly
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.InvocationContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.SessionContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererPopupManager
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererUserActionListener
+import software.aws.toolkits.jetbrains.services.codewhisperer.util.CodeWhispererConstants.TOTAL_SECONDS_IN_MINUTE
+import software.aws.toolkits.telemetry.CodewhispererLanguage
+import software.aws.toolkits.telemetry.CodewhispererTelemetry
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.roundToInt
+
+abstract class CodeWhispererCodeCoverageTracker(
+    private val timeWindowInSec: Long,
+    private val language: CodewhispererLanguage,
+    acceptedTokensSize: AtomicInteger,
+    totalTokensSize: AtomicInteger
+) : Disposable {
+    val percentage: Int
+        get() = if (totalTokensSize.get() != 0) (acceptedTokensSize.get().toDouble() / totalTokensSize.get() * 100).roundToInt() else 0
+    var acceptedTokensSize = acceptedTokensSize
+        private set
+    var totalTokensSize = totalTokensSize
+        private set
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+    private val isShuttingDown = AtomicBoolean(false)
+    private var startTime: Instant = Instant.now()
+
+    init {
+        val conn = ApplicationManager.getApplication().messageBus.connect()
+        conn.subscribe(
+            CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+            object : CodeWhispererUserActionListener {
+                override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, remainingRecomm: String) {
+                    if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
+                    addAndGetAcceptedTokens(remainingRecomm.length)
+                }
+            }
+        )
+        scheduleCodeWhispererCodeCoverageTracker()
+    }
+
+    fun documentChanged(event: DocumentEvent) {
+        addAndGetTotalTokens(event.newLength - event.oldLength)
+    }
```

Comment:
If we just need the document length every 5 mins, can we just get the length when we need it instead of keeping this val? Do we need intermediate total length value for other purposes?

Reply:
I dont get this one @andrewyuq , could you elaborate on it a bit thanks!

Reply:
Nvm I understanded this wrong, it's aggregated per language.